[FINNA-3151] Implement IIIF support with Tify

ronja-koistinen opened the pull request:
This PR is another alternative to:

#3316
#3358 
#3364

...and so far my personal favourite.

* IIIF metadata trait for RecordDriver
  * Finna\RecordDriver\SolrMarc: method returning IIIF manifests

* UI component template using Tify as the viewer
  * Displayed on the record page

* UI component template for IIIF tools dropdown menu
  * Displayed on the record page and record-image-paginator

ronja-koistinen committed:
IIIF functionality

* IIIF metadata trait for RecordDriver
  * Finna\RecordDriver\SolrMarc: method returning IIIF manifests

* UI component template using Tify as the viewer
  * Displayed on the record page

* UI component template for IIIF tools dropdown menu
  * Displayed on the record page and record-image-paginator

ronja-koistinen committed:
Style fixes

ronja-koistinen commented:
Tyylit pitää vielä mukauttaa vastaamaan Heidin tekemiä leiskoja, kuten Cloverissa jo tein, mutta se on varmaan päivän homma.

EreMaijala commented on module/Finna/src/Finna/RecordDriver/Feature/FinnaIiifTrait.php:
```diff
@@ -0,0 +1,66 @@
+     *
+     * @return array
+     */
+    public function getAllIiifManifests(): array
```
Oletuksena kaikki jutut palauttaa kaikki, niin "All" ei ole tarpeen.
```suggestion
    public function getIiifManifests(): array
```

EreMaijala commented on docker-compose.yml:
```diff
@@ -81,7 +81,8 @@ services:
       - type: volume
         source: finna-mariadb
-        target: /var/lib/mysql:rw
+        target: /var/lib/mysql
```
Voisko tämä olla erillinen PR?

ronja-koistinen replied:
Otin pois.

EreMaijala commented on themes/finna2/templates/_ui/components/finna-iiif-menu.phtml:
```diff
@@ -0,0 +1,60 @@
+    </span>
+  </li>
+  <li>
+    <a class="dropdown-item btn toolbar-btn" href="<?=$this->manifestURL?>"
```
Kuvakkeellisella linkillä pitäisi olla luokka `icon-link`, sen ikonilla `icon-link__icon` ja labelilla `icon-link__label`.

EreMaijala commented on themes/finna2/templates/_ui/components/finna-iiif-viewer.phtml:
```diff
@@ -0,0 +1,57 @@
+</div>
+
+<div class="record-iiif-toolbar container">
+    <span class="record-iiif-functions dropdown hstack gap-3">
```
Templaateissa yleensä sisennys kaksi välilyöntiä.

EreMaijala commented on themes/finna2/templates/_ui/components/finna-iiif-menu.phtml:
```diff
@@ -0,0 +1,60 @@
+$r = $this->alignRight ? 'dropdown-menu-end' : '';
+$urlContainerId = 'iiif-manifest-url__' . ($this->uniqueID ?? '');
+?>
+<span class="iiif-manifest-url-container"
```
```suggestion
<span class="iiif-manifest-url-container hidden"
```

EreMaijala commented on themes/finna2/templates/_ui/components/finna-iiif-menu.phtml:
```diff
@@ -0,0 +1,60 @@
+<span class="iiif-manifest-url-container"
+      id="<?=$urlContainerId?>"
+      aria-hidden="true">
+  <?=$this->manifestURL?>
```
escapeHtml?

ronja-koistinen committed:
Address review comments

ronja-koistinen committed:
Override iiif-parser version

In an attempt to address the CodeQL complaint https://github.com/IIIF-Commons/parser/pull/57

ronja-koistinen committed:
Fix typo in method name

ronja-koistinen commented:
Tässä nyt vielä on tuo CodeQL-varoitus, joka liittyy ainoastaan siihen, että `@iiif/parser` validoi, onko manifestien `license`-kentässä sallitunlaisia urleja. Ei liity mihinkään oikeaan pyyntöjen validointiin.

Lisäksi tämä toteutus toistaiseksi lävähtää saman tien auki suoraan tietuesivulle, mutta lopullinen suunnitelma on ilmeisesti avata IIIF-katselu lightboxiin, jota ei ole vielä toteutettu.

ronja-koistinen committed:
Probably better wording in Sámi

ronja-koistinen committed:
Translation changes from feedback

ronja-koistinen committed:
Tify theme changes

ronja-koistinen committed:
CSS updates

ronja-koistinen committed:
Update sass to 1.94.2

This is to gain access to the $method argument in color.mix()

ronja-koistinen committed:
Various fixes

* Lighter TIFY background colors
* Remove unnecessary border
* CSS comments
* Force two-letter ISO 639 language code for TIFY l10n

ronja-koistinen commented:
Aiemmin mainittu lightbox-homma ei koskekaan tietuesivua vaan hakutulossivua; tähän palataan myöhemmin.

Tifystä puuttuu nyt vielä suomen-, ruotsin- ja saamenkieliset käyttöliittymätekstit. Suomenkieliset on olemassa ja [odottamassa mergeä yläjuoksulla](https://github.com/tify-iiif-viewer/tify/pull/272).

IIIF-parserikirjaston versio, joka toivottavasti ei enää hälytä CodeQL:ssä, on [odottamassa päivitystä yläjuoksulla](https://github.com/tify-iiif-viewer/tify/pull/269).

[Muuta customointia koskeva keskustelu-issue avoinna myös.](https://github.com/tify-iiif-viewer/tify/issues/275)

ronja-koistinen commented:
Ai niin paitsi se Lataa kuva -nappi ei vielä tee mitään...

ronja-koistinen committed:
Media controls: improve contrast, move to right-hand side

ronja-koistinen committed:
Remove "Download image" button

Equivalent functionality available in TIFY's "Export" menu

ronja-koistinen committed:
Disable thumbnail when in IIIF mode

ronja-koistinen committed:
Revert move of media controls to right-hand side

This was a misunderstanding of agreed plans.

EreMaijala commented on package.json:
```diff
@@ -29,7 +29,7 @@
     "grunt-dart-sass": "^2.0.1",
     "jit-grunt": "^0.10.0",
-    "sass": "1.72.0"
+    "sass": "1.94.2"
```
Jos tämä tarvitaan, niin pitää laittaa yläjuoksulle.

EreMaijala commented on themes/finna2/templates/RecordDriver/DefaultRecord/record-image-paginator.phtml:
```diff
@@ -86,8 +92,23 @@
+         alt="<?=$this->transEsc('No Cover Image')?>">
   </a>
+  <?php if ($manifests): ?>
+  <button type="button"
```
Sisennä if:n sisällä oleva osa.

EreMaijala commented on themes/finna2/templates/RecordDriver/DefaultRecord/core.phtml:
```diff
@@ -60,7 +61,9 @@
   <div class="media-<?=$thumbnailAlignment ?><?=!empty($audioUrls) ? ' audio' : ''?>">
 
-    <?=$this->record($this->driver)->renderTemplate('record-image-information.phtml', ['img' => $img]);?>
+    <?php if ($iiifManifests): ?>
```
Mikä tämän tarkoitus on? Nyt tavalliset kansikuvat eivät näy.

ronja-koistinen replied:
En oikeastaan ihan muista. Sen varmaan piti olla `if (!$iiifManifests)`, koska ajatus oli varmaan, että tietuesivulla on outoa, jos näkyy sekä perinteinen thumbnail (josta aukeaa isompi kuva modaaliin) että IIIF-katselu.

ronja-koistinen replied:
Nyt se on `if (!$iiifManifests)`

EreMaijala commented on themes/finna2/templates/RecordDriver/DefaultRecord/record-image-paginator.phtml:
```diff
@@ -67,11 +67,17 @@
 ?>
 
-<div class="recordcover-container">
+<div class="recordcover-container position-relative">
```
Vaikka meillä tuollainen onkin, niin tykkäisin ennemmin, että .recordcover-container sisältäisi tarpeellisen positionin.

ronja-koistinen replied:
En muista edes, miksi olen ton lisännyt. Tuon luokan poistaminen ei vaikuta muuttavan näkymää mitenkään. Otan pois.

EreMaijala replied:
Jos siellä on jotain absoluuttisesti positioitua, niin voi tarvita tolle containerille `position: relative`:n. Eli tulee tyypillisesti esiin jonkun valikon tms. kohdalla.

EreMaijala replied:
Ja jos se tosiaan olisi tarpeen, niin mahdollista myöhempää ihmettelyä helpottaa, jos laittaa scss:ään pienen kommentin siitä, miksi se on siellä.

ronja-koistinen replied:
En oikeasti nyt muista, miksi laitoin siihen tuon. En näe nyt mitään tarvetta, kaikki toimii ilmankin.

EreMaijala commented on themes/finna2/templates/RecordDriver/DefaultRecord/record-image-paginator.phtml:
```diff
@@ -67,11 +67,17 @@
+     href="<?=$this->escapeHtmlAttr($this->recordLinker()->getUrl($this->driver)) ?>"
+     data-images="<?=htmlspecialchars(json_encode($images ?? []), ENT_QUOTES, 'UTF-8');?>"
+     data-settings="<?=htmlspecialchars(json_encode($settings ?? []), ENT_QUOTES, 'UTF-8');?>"
+     tabindex="-1">
```
Koodin luettavuuden parantamiseksi tehtävät muutokset mieluummin erikseen, muuten tulee vaikea diffi. Ja jos sille tielle lähtee, niin käyttöön htmlAttributes-helpperi.

ronja-koistinen replied:
Peruin nämä mielestäni entiselleen.

EreMaijala commented on themes/finna2/templates/_ui/components/finna-iiif-menu.phtml:
```diff
@@ -0,0 +1,62 @@
+  </li>
+  <li>
+    <a class="dropdown-item btn toolbar-btn"
+      href="https://uv-v4.netlify.app/#?manifest=<?=$encodedManifest?>"
```
Ulkoiset viewerit pitäisi konffata config.ini:ssä, ettei tarvitse koodia muuttaa niiden päivittämiseksi, lisäämiseksi tai poistamiseksi. Esim. samaan tapaan kuin [LinkPopovers]-sektiossa `external_links` (toki ilman regex:iä, placeholder vaan manifestille).

ronja-koistinen replied:
Konffiavaimet on nyt.

EreMaijala commented on themes/finna2/templates/_ui/components/finna-iiif-viewer.phtml:
```diff
@@ -0,0 +1,57 @@
+$cssPath = $this->scriptSrc('vendor/tify/tify.css');
+?>
+
+<link rel="stylesheet" href="<?= $cssPath; ?>">
```
Tämä pitäisi lisätä assetManagerilla. Mutta se sitten aiheuttaisi toisen bundlen luomisen. Sanoisin, että voitaisiin ottaa tämä finna2-teemaan vakituiseksi, eli lisätään theme.config.php:hen. Pitäisi vaan laittaa ennemmin css/vendor:n alle, ja nimeksi tify.min.css, koska se on valmiiksi minifioitu, eikä haluta minifioida kahteen kertaan.

EreMaijala commented on themes/finna2/templates/_ui/components/finna-iiif-viewer.phtml:
```diff
@@ -0,0 +1,57 @@
+
+<?php
+$tifyJsPath = $this->scriptSrc('vendor/tify/tify.js');
+$tifyJsPathEsc = addslashes($tifyJsPath);
```
Käyttäisin json_encode:a ennemmin. Ja oikeastaan koko tuon config-objektin voisi tehdä arrayna ja json_encode:ttaa sitten tuohon JS:ään.

ronja-koistinen replied:
Se `addslashes()` on sitä varten, jos siellä stringissä on jostain syystä lainausmerkkejä. Ja sehän pitää joka tapauksessa olla, jos tässä jsonia käytetään, koska sit JS:ssä pitää olla:

```javascript
new Tify(JSON.parse("$tifyConfig"));
```

joka hajoaa, jos `$tifyConfig` on json_encoden läpi vedetty PHP-objekti, koska siellä on lainausmerkkejä väistämättä. Joko `"` tai `'`.

ronja-koistinen replied:
Se toimii nyt näin:

```php
$tifyJsPath = $this->scriptSrc('vendor/tify/tify.js');
$tifyJsPathEsc = addslashes($tifyJsPath);
$tifyConfig = json_encode([
    'container' => '#tify',
    'manifestUrl' => $this->manifestUrls[0],
    'language' => substr($this->language, 0, 2) ?? 'fi',
    'colorMode' => 'light',
]);
$js = <<<JS
    import Tify from '$tifyJsPathEsc';
    new Tify(JSON.parse('$tifyConfig'));
    JS;
```

mutta lakkaa heti toimimasta, kun `$tifyConfig` slsältää jostain syystä `'`-merkin, jos en lisää sitä addslashes-vaihetta vielä.

EreMaijala replied:
Ei sitä json_encode:n tulosta tartte parsia, se on valmis objekti JSON-muodossa, eli sen voi antaa suoraan:
`new Tify($tifyConfig);`

EreMaijala replied:
Ks. esim. themes/bootstrap5/templates/Helpers/cookie-consent.phtml

ronja-koistinen replied:
Nyt on toteutus json_encodella

EreMaijala commented on themes/finna2/templates/record/view.phtml:
```diff
@@ -120,6 +120,13 @@
     <?php endif; ?>
   <?php endif; ?>
+  <?php $iiifManifests = $this->driver->tryMethod('getIiifManifests');
+    if ($iiifManifests):
```
```suggestion
  <?php 
    if ($iiifManifests = $this->driver->tryMethod('getIiifManifests')):
```

EreMaijala commented on module/Finna/src/Finna/RecordDriver/Feature/FinnaIiifTrait.php:
```diff
@@ -0,0 +1,66 @@
+ *
+ * PHP version 8
+ *
+ * Copyright (C) The National Library 2015-2025.
```
Ei kai sentään 2015 asti ole tätä yritetty?

EreMaijala commented on themes/finna2/scss/finna/iiif.scss:
```diff
@@ -0,0 +1,134 @@
+}
+
+.iiif-manifest-url-container {
+    display: none;
```
Turha, kun sillä on hidden-luokka kuitenkin?

EreMaijala commented on themes/finna2/templates/RecordDriver/DefaultRecord/record-iiif-viewer.phtml:
```diff
@@ -0,0 +1,4 @@
+<!-- START of: finna - RecordDriver/DefaultRecord/iiif-viewer.phtml -->
+<?=$this->component('finna-iiif-viewer',
+    ['manifestURL' => $manifests[0], 'language' => $this->layout()->userLang]);?>
```
Voisiko komponentti kaivaa kielen itse, turha kai sitä on välittää?

EreMaijala replied:
Komponentille voisi välittää koko manifestilistan. Se voi nyt näyttää siitä ensimmäisen, mutta olisi sitten rakenne valmiina, jos halutaan jossain vaiheessa antaa käyttäjän valita.

EreMaijala commented on module/Finna/src/Finna/RecordDriver/SolrMarc.php:
```diff
@@ -165,6 +166,31 @@ public function getAccessRestrictionsType($language)
+                $manifests[] = $u;
+            }
+        }
+        return $manifests;
```
Muuttaisin tätä sen verran, että palauttaa valmiiksi assosiatiivisen arrayn:
```
[
  ['url' => 'https:/...'],
  ['url' => 'https:/...'],
]  
```
Ei sitten tule ongelmia, jos tuohon jossain vaiheessa halutaan lisää tietoa esim. manifestin valitsemista varten.

ronja-koistinen committed:
Revert "Update sass to 1.94.2"

This reverts commit b32da942dad1411d3540ef1c95051733c97ac099.

EreMaijala commented on themes/finna2/templates/_ui/components/finna-iiif-viewer.phtml:
```diff
@@ -0,0 +1,57 @@
+  <span class="record-iiif-functions dropdown hstack gap-3">
+    <button type="button" class="btn btn-outline-primary p-2 ms-auto icon-link" data-bs-toggle="dropdown"
+      aria-expanded="false">
+      <img src="<?=$this->imageSrc()->getSourceAddress('iiif_logo.min')?>" alt="IIIF" class="bi">
```
src pitää escapeta (escapeHtmlAttr tai escapeHtml).

ronja-koistinen committed:
Use flat gray bg for TIFY, easier to do

ronja-koistinen committed:
TIFY installation changes

* Only install tify.js and tify.css
  * Exclude stuff for the demo site
* Rename to tify.min.js and tify.min.css
* Include tify.min.css in theme.config.php
* Refactor copyDependencies.sh

ronja-koistinen committed:
Implement suggested changes from review + more

* Update some doc comments
* Add more escapeHtml
* Add config keys for external IIIF viewers

ronja-koistinen committed:
Fix whitespace stuff

ronja-koistinen committed:
Include Tify translation files

ronja-koistinen commented on themes/finna2/package.json:
```diff
@@ -0,0 +1,16 @@
+  },
+  "overrides": {
+    "@iiif/parser": ">=2.2.6"
+  }
```
Tämän voisi varmaan ottaa pois, koska ei tainnut toimia halutulla tavalla.

ronja-koistinen commented on themes/finna2/templates/_ui/components/finna-iiif-menu.phtml:
```diff
@@ -0,0 +1,68 @@
+$encodedManifest = urlencode($manifest);
+$r = $this->alignRight ? 'dropdown-menu-end' : '';
+$urlContainerId = 'iiif-manifest-url__' . ($this->uniqueID ?? 'container');
+$viewers = $this->config()->get('config')->IiifExternalViewers->viewers ?? [];
```
Vai olisiko sittenkin parempi `IIIF->externalViewers`?

EreMaijala replied:
Joo, olisi. Sitten [IIIF]-sektioon voisi tarvittaessa laittaa muitakin asetuksia.

EreMaijala commented on module/Finna/src/Finna/View/Helper/Root/ScriptSrc.php:
```diff
@@ -49,7 +49,7 @@ class ScriptSrc extends ThemeSrc
     public function __invoke($source)
     {
-        if ($url = $this->fileFromCurrentTheme("js/$source")) {
+        if ($url = $this->fileFromCurrentTheme("js/$source", false, true)) {
```
```suggestion
        if ($url = $this->fileFromCurrentTheme("js/$source", allowParentThemes: true)) {
```

EreMaijala commented on module/Finna/src/Finna/RecordDriver/SolrMarc.php:
```diff
@@ -165,6 +167,31 @@ public function getAccessRestrictionsType($language)
+     *
+     * @return array
+     */
+    public function getIiifManifests()
```
```suggestion
    public function getIiifManifests(): array
```

EreMaijala commented on themes/finna2/scss/finna/iiif.scss:
```diff
@@ -0,0 +1,121 @@
+    }
+
+    @media screen and (width <= 700px) {
+        & {
```
Turha rivi?

EreMaijala commented on themes/finna2/scss/finna/iiif.scss:
```diff
@@ -0,0 +1,121 @@
+.iiif-menu-button-badge {
+    @extend .btn, .btn-secondary;
+
+    @media screen and (width > 700px) {
```
Onko tuo 700 jotenkin tähän käyttöön tarkka arvo, vai voisiko käyttää valmiita `media-breakpoint-up`/`media-breakpoint-down` helpereitä kuten muualla? sm-breakpoint olisi 768px.

EreMaijala commented on themes/finna2/scss/finna/record.scss:
```diff
@@ -23,6 +23,10 @@
   }
+
+  .media-left:empty,
+  .media-right:empty { display: none }
```
Tarvitaanko tätä? media-left tai media-right jätetään pois, jos on manifesti, eikö?

EreMaijala commented on themes/finna2/templates/RecordDriver/DefaultRecord/core.phtml:
```diff
@@ -60,7 +61,9 @@
   ob_start(); ?>
```
Jätä $thumbnail rendaamatta, jos on $iiifManifests. Sitten ei tarvitse sitä enää tarkistella siellä, missä thumbnailia käytetään, eikä rendata turhaan jotain, mitä ei käytetä.

EreMaijala commented on themes/finna2/templates/_ui/components/finna-iiif-menu.phtml:
```diff
@@ -0,0 +1,68 @@
+    <?php
+    foreach ($viewers as $viewer):
+        $viewerName = strtok($viewer, '|');
+        $viewerHref = strtok('|');
```
```suggestion
        [$viewerName, $viewerHref] = explode('|', "$viewer|");
```

EreMaijala commented on themes/finna2/templates/_ui/components/finna-iiif-menu.phtml:
```diff
@@ -0,0 +1,68 @@
+        $viewerHref = str_replace('{manifest}', $encodedManifest, $viewerHref);
+        ?>
+        <li>
+            <a class="dropdown-item btn toolbar-btn"
```
a:lla pitää olla icon-link -luokka, iconilla icon-link__Icon ja tekstillä icon-link__label kuten ylempänäkin.

EreMaijala commented on themes/finna2/templates/_ui/components/finna-iiif-viewer.phtml:
```diff
@@ -0,0 +1,50 @@
+            aria-expanded="false">
+      <img src="<?=$this->escapeHtmlAttr($iiifLogo)?>" alt="IIIF" class="bi">
+      <?=$this->transEsc('iiif_tools')?>
+      <?=$this->icon('angle-down')?>
```
```suggestion
      <?=$this->icon('dropdown-open')?>
```

ronja-koistinen committed:
Changes from review

ronja-koistinen committed:
Change IIIF section in config.ini.sample

EreMaijala commented on themes/finna2/templates/RecordDriver/DefaultRecord/core.phtml:
```diff
@@ -56,29 +57,31 @@
-    <?php if (!empty($videoUrls)): ?>
+      <?=$this->record($this->driver)->renderTemplate('record-image-information.phtml', ['img' => $img]);?>
+      <?php if (!empty($videoUrls)): ?>
       <?=$this->record($this->driver)->renderTemplate('record-video-player.phtml', ['videoUrls' => $videoUrls]);?>
```
Nämä if:n sisällä olevatkin pitäisi sisentää.

EreMaijala commented on themes/finna2/templates/RecordDriver/DefaultRecord/record-image-paginator.phtml:
```diff
@@ -88,6 +90,20 @@
   </a>
+  <?php if ($manifests): ?>
+    <button type="button"
+            class="badge text-bg-light position-absolute bottom-0 end-0
```
Laita kaikki luokat yhdelle riville. Ei se haittaa templaatissa, vaikka tulisi vähän pidempi. Tai käytä htmlAttributes-helperiä. Silloin voi laittaa kaikki attribuutit array:hin ensin.

EreMaijala commented on themes/finna2/templates/_ui/components/finna-iiif-menu.phtml:
```diff
@@ -0,0 +1,69 @@
+  </li>
+  <?php if ($viewers):?>
+    <li>
+        <span class="dropdown-header fs-5">
```
Tässä muuttuu sisennys yhtäkkiä kahdesta neljään, käytä kahta koko tiedostossa.

EreMaijala reviewed (changes requested):
Odotellaanko muuten vielä suomenkielisiä käännöksiä tms. ennen mergeä, vai otetaanko ne sitten myöhemmin?

ronja-koistinen committed:
Suggestions from review

ronja-koistinen commented:
> Odotellaanko muuten vielä suomenkielisiä käännöksiä tms. ennen mergeä, vai otetaanko ne sitten myöhemmin?

Suomennos on mergetty Tifyn yläjuoksulle tällä viikolla, mutta uutta versiota ei ole vielä julkaistu Githubissa tai NPM:ssä, eli sitä varmaan joutuu vähän vielä odottamaan.
